File: archive-application/src/main/java/site/archive/service/user/UserService.java
package site.archive.service.user;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import site.archive.common.exception.common.DuplicateFieldValueException;
import site.archive.common.exception.common.ResourceNotFoundException;
import site.archive.domain.user.OAuthUserRepository;
import site.archive.domain.user.PasswordUserRepository;
import site.archive.domain.user.UserRepository;
import site.archive.dto.v1.user.BaseUserDtoV1;
import site.archive.dto.v1.user.SpecificUserDtoV1;

@Service
@Transactional(readOnly = true)
@Slf4j
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordUserRepository passwordUserRepository;
    private final OAuthUserRepository oAuthUserRepository;

    public BaseUserDtoV1 findUserById(long userId) {
        return userRepository.findById(userId)
                             .map(BaseUserDtoV1::from)
                             .orElseThrow(() -> new ResourceNotFoundException("아이디에 해당하는 유저가 존재하지 않습니다."));
    }

    public BaseUserDtoV1 findSpecificUserById(long userId) {
        var user = findUserById(userId);
        updateUserTypeWithProviderWhenOAuthUser(userId, user);
        return user;
    }

    public boolean existsEmail(String email) {
        return userRepository.findByMailAddress(email).isPresent();
    }

    public boolean existsNickname(String nickname) {
        return userRepository.findByNickname(nickname).isPresent();
    }

    @Transactional
    public void deleteUser(long userId) {
        userRepository.deleteById(userId);
    }

    @Transactional
    public void updateUserNickname(long userId, String nickname) {
        if (existsNickname(nickname)) {
            throw new DuplicateFieldValueException("nickname", nickname);
        }
        userRepository.updateNickName(userId, nickname);
    }

}

File: archive-application/src/main/kotlin/site/archive/service/user/UserService.kt
package site.archive.service.user

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import site.archive.common.exception.common.DuplicateFieldValueException
import site.archive.common.exception.common.ResourceNotFoundException
import site.archive.domain.user.UserRepository
import site.archive.dto.v1.user.BaseUserDtoV1
import site.archive.dto.v1.user.SpecificUserDtoV1

@Service
@Transactional(readOnly = true)
class UserService(val userRepository: UserRepository) {

    private val log = LoggerFactory.getLogger(javaClass)

    fun findUserById(userId: Long): BaseUserDtoV1 {
        return userRepository.findById(userId)
            .map(BaseUserDtoV1::from)
            .orElseThrow { ResourceNotFoundException("아이디에 해당하는 유저가 존재하지 않습니다.") }
    }

    fun findUserByEmail(email: String): BaseUserDtoV1 {
        return userRepository.findByMailAddress(email)
            .map(BaseUserDtoV1::from)
            .orElseThrow { ResourceNotFoundException("가입되지 않은 Email 입니다.") }
    }

    fun findSpecificUserById(userId: Long): SpecificUserDtoV1 {
        val user = userRepository.findById(userId)
            .orElseThrow { ResourceNotFoundException("아이디에 해당하는 유저가 존재하지 않습니다.") }
        return SpecificUserDtoV1.from(user)
    }

    fun existsEmail(email: String): Boolean {
        return userRepository.findByMailAddress(email).isPresent
    }

    fun existsNickname(nickname: String): Boolean {
        return userRepository.findByNickname(nickname).isPresent
    }

    @Transactional
    fun deleteUser(userId: Long) {
        log.info("User($userId)이 탈퇴했습니다")
        userRepository.deleteById(userId)
    }

    @Transactional
    fun updateUserNickname(userId: Long, nickname: String) {
        if (existsNickname(nickname)) {
            throw DuplicateFieldValueException("nickname", nickname)
        }
        userRepository.updateNickName(userId, nickname)
    }

}